MFDecodeH264/MFDecodeH264.cpp

```diff
@@ -1,147 +1,581 @@
+
+        // So we don't know enough information yet, we'll feed input samples until we get MF_E_TRANSFORM_STREAM_CHANGE and then we'll provide a sample as per doc:
+        //   "If the input type contains only these two attributes, the decoder will offer a default output type, which acts as a placeholder."
+        //   "When the decoder receives enough input samples to produce an output frame, it signals a format change by returning MF_E_TRANSFORM_STREAM_CHANGE"
+        UINT32 sampleSize = 0;
+
+        // input type is H264
+        winrt::com_ptr<IMFMediaType> inputType;
+        HRCHECK(MFCreateMediaType(inputType.put()));
+        HRCHECK(inputType->SetGUID(MF_MT_MAJOR_TYPE, MFMediaType_Video));
+        HRCHECK(inputType->SetGUID(MF_MT_SUBTYPE, MFVideoFormat_H264));
+        HRCHECK(inputType->SetUINT32(MF_MT_ALL_SAMPLES_INDEPENDENT, TRUE));
+        HRCHECK(inputType->SetUINT32(MF_MT_FIXED_SIZE_SAMPLES, TRUE));
+
+        HRCHECK(decoder->SetInputType(0, inputType.get(), 0)); // video is id 0
+        // get (and set) NV12 output type (could be I420, IYUV, YUY2, YV12)
+        HRCHECK(SetOutputType(decoder, MFVideoFormat_NV12));    
+        int notAccept = 0;
+        auto file = CreateFile(L"h264.h264", GENERIC_READ, 0, nullptr, OPEN_EXISTING, 0, nullptr);
+        WIN32CHECK(file != INVALID_HANDLE_VALUE);
+
+        DWORD TotalRead = 0;
+        for(bool Done = false; !Done; )
+        {
+            // get a random chunk size between 500 and 1500
+            DWORD chunkSize = 500 + (1000 * (RAND_MAX - std::rand())) / RAND_MAX;
+
+            // create an MF input buffer & read into it
+            winrt::com_ptr<IMFMediaBuffer> inputBuffer;
+            HRCHECK(MFCreateMemoryBuffer(chunkSize, inputBuffer.put()));
+            BYTE* chunk;
+            HRCHECK(inputBuffer->Lock(&chunk, nullptr, nullptr));
+            DWORD read;
+            WIN32CHECK(ReadFile(file, chunk, chunkSize, &read, nullptr));
+            HRCHECK(inputBuffer->SetCurrentLength(read));
+            HRCHECK(inputBuffer->Unlock());
+            if (read)
+            {
+                winrt::com_ptr<IMFSample> inputSample;
+                HRCHECK(MFCreateSample(inputSample.put()));
+                HRCHECK(inputSample->AddBuffer(inputBuffer.get()));
+
+                auto hr = decoder->ProcessInput(0, inputSample.get(), 0);
+                if (hr != MF_E_NOTACCEPTING) // just go on
+                {
+                    HRCHECK(hr);
+                }
+
+                if (hr == MF_E_NOTACCEPTING) {
+                    OutputDebugStringA("Not accepting!\n");
+                    notAccept++;
+                }
+            }
+            else
+            {
+                _RPTWN(_CRT_WARN, L"Draining, TotalRead %u\n", TotalRead);
+                // end of file, ask decoder to process all data from previous calls
+                HRCHECK(decoder->ProcessMessage(MFT_MESSAGE_COMMAND_DRAIN, 0));

```

Comment:
You drain here and it is safe to expect that you will have full output before the MFT gives "need input" again.

I think there might be a bug in MFT implementation which discards some of the data, but this might also happen because - after all - you are normally expected to feed ES input frame by frame to the decoder, and in your sample you separate H.264 input into "packets" randomly, not at frame boudnary. 

Hence, the effect.